### ISSUE TITLE
jsii doesn't work with pnpm workspace

### ISSUE THREAD

**Issue by chengjianhua** [aws/jsii#4016]
### Describe the feature

At present, jsii supports 2 kinds of dependency version resolving:
1. General semver version like `^1.1.0`
2. File dependency like `file:../../some-local-package`

Propose jsii to support Pnpm workspace protocol version, like `workspace:*`.

```json5
// packages/pkg-a/package.json
{
  "dependencies": {
    "some-package": "workspace:*"
  }
}
```

```
- root (with Pnpm Workspace enabled)
	- packages
		- package-a
		- some-package
```

### Use Case

I tried to use jsii under a rush x pnpm workspace monorepo. When I ran `jsii` in a sub-package that has dependencies like `"some-package": "workspace:*"`.

Then `jsii` will throw the error `Invalid semver expression for some-package: workspace:*`. The relevant codes are located in https://github.com/aws/jsii/blob/ae16abc57bdd7fd439c4183cfcb015acadaaa3ae/packages/jsii/lib/project-info.ts#L355

The current behavior on version resolving doesn't work with pnpm workspace. This stops me from using jsii under pnpm workspace.

* https://pnpm.io/workspaces

### Proposed Solution

I tried tampering with some codes in `project-info.ts`, then jsii works for me.

1. add the workspace protocol check in `_resolveVersion`
```js
    const isWorkspace = /^workspace:/.test(dep);
    if (isWorkspace) {
        const dependencyDir = utils_1.findDependencyDirectory(mod, searchPath);
        console.log('found dependencyDir', dependencyDir);
        const version = fs.readJsonSync(path.join(dependencyDir, 'package.json')).version;
        return {
            version: version,
        }
    }
```

2. ignore semver check in `resolveDependency` for `workspace:*` version

3. ignore the comparison between the actual version and declaration in `discoverDependencyTree`

After making the above changes, `jsii` will not throw errors related to dependency resolving again.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.69.0

### Environment details (OS name and version, etc.)

os: MacOS jsii: 1.78.1

**Comment by chengjianhua** [comment]
@RomainMuller Could you help take a look at this feature request?

**Comment by zsstiers** [comment]
Is this issue something that a PR would be accepted for? I just ran into this same issue and discovered this ticket that is nearly a year old.

@chengjianhua, if you still have your fix maybe just posting a PR would be more successful.

**Comment by mrgrain** [comment]
Yes, we would accept a PR for this. I think jsii should accept any version constraint that is not a semver version, but warn the user about the fact that it cannot check the correctness of the version and leave it up to them to proceed.
